### PR TITLE
Multiview pipeline

### PR DESCRIPTION
**Pull request by dkwingsmt**
This PR makes `Animator` able to handle multiple views, and updates unit tests accordingly.

Before:
image

After:
image

Now `Animator::Render` must be called during `Animator::BeginFrame`, which is split into `BeginFrame` and `EndFrame`. This requirement is made possible by https://github.com/flutter/engine/pull/45555. The reason to split is to allow `ShellTest::PumpOneFrame` to insert a render from C++ code.

`ShellTest::PumpOneFrame` is also refactored to allow pumping a frame without any views.

A few unit tests are tweaked to resolve racing condition.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by dkwingsmt** [comment]
@ColdPaleLight While developing this PR I also encountered an extremely rare flake for test https://github.com/flutter/engine/pull/44473/files#diff-b5ee07fa281c3906cecc4f4da35b074a58ceba6259f0f3869c4d9cab97b9f821R2086 , where this test appears to call the secondary callback first and the primary callback second.

```
[92/350] ShellTest.SecondaryVsyncCallbackShouldBeCalledAfterVsyncCallback (4192 ms)
[INFO:test_timeout_listener.cc(76)] Test timeout of 300 seconds per test case will be enforced.
[0;33mNote: Google Test filter = ShellTest.SecondaryVsyncCallbackShouldBeCalledAfterVsyncCallback
[m[0;32m[==========] [mRunning 1 test from 1 test suite.
[0;32m[----------] [mGlobal test environment set-up.
[0;32m[----------] [m1 test from ShellTest
[0;33m[ DISABLED ] [mShellTest.DISABLED_SkipAndSubmitFrame
[0;32m[ RUN      ] [mShellTest.SecondaryVsyncCallbackShouldBeCalledAfterVsyncCallback
../../flutter/shell/common/shell_unittests.cc:2113: Failure
Value of: is_on_begin_frame_called
  Actual: false
Expected: true
../../flutter/shell/common/shell_unittests.cc:2096: Failure
Value of: is_secondary_callback_called
  Actual: true
Expected: false
[0;31m[  FAILED  ] [mShellTest.SecondaryVsyncCallbackShouldBeCalledAfterVsyncCallback (3937 ms)
```
https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8767758920767307281/+/u/test:_test:_Host_Tests_for_host_debug_unopt/stdout?format=raw

I'm not really sure what the cause was, but I think it's not related to this PR. Nevertheless, just for your information.

File: flow/frame_timings.cc [hunk 1]
```diff
@@ -255,7 +255,9 @@ const char* FrameTimingsRecorder::GetFrameNumberTraceArg() const {
 }
 
 void FrameTimingsRecorder::AssertInState(State state) const {
-  FML_DCHECK(state_ == state);
+  FML_DCHECK(state_ == state)
+      << "Expected state " << static_cast<uint32_t>(state) << ", actual state "
+      << static_cast<uint32_t>(state_);
```
Comment (dnfield) [inline flow/frame_timings.cc:260]: nit: Consider a (debug-only) operator to convert the state to a friendly string.

File: shell/common/animator.cc [hunk 2]
```diff
@@ -112,6 +116,33 @@ void Animator::BeginFrame(
...
+    delegate_.OnAnimatorUpdateLatestFrameTargetTime(
+        frame_timings_recorder_->GetVsyncTargetTime());
+
+    // Commit the pending continuation.
+    PipelineProduceResult result =
+        producer_continuation_.Complete(std::make_unique<FrameItem>(
+            std::move(layer_trees_tasks_), std::move(frame_timings_recorder_)));
+
+    if (!result.success) {
+      FML_DLOG(INFO) << "Failed to commit to the pipeline";
```
Comment (dnfield) [inline shell/common/animator.cc:136]: This falls through where it used to return early. 

Do we really want to continue below in this case?

Comment (dkwingsmt) [reply shell/common/animator.cc:136]: When I was developing this PR I had the same question, but I think the current way is keeping the old behavior.

Before the PR, although when `result.success` is false, `Animator::Render` returned early, it was all contained in `delegate_.OnAnimatorBeginFrame`, while the `if (!frame_scheduled_ && has_rendered_)` clause continued to be executed.

If this logic seems questionable, I can open an issue to remind further checking.

File: flow/frame_timings.cc (hunk 1 above)
Comment (dkwingsmt) [reply flow/frame_timings.cc]: Done.

**Comment by dnfield** [comment]
This LGTM but I'll look at it again once it's up to date.

**Comment by chinmaygarde** [comment]
xref the re-land https://github.com/flutter/engine/pull/47095

**Comment by dkwingsmt** [comment]
The code that this PR relies on has been relanded in https://github.com/flutter/engine/pull/47062, which I've checked has no performance regressions. @dnfield Can you take a final look?

**Review by dnfield — APPROVED** [review]

**Pull request merged**